p2p/src/peer.rs

```diff
@@ -1,657 +1,653 @@
+                let _ = terminated_sender.send(Terminated { conn_id, peer_id });
+                return;
             }
-            Peer::Disconnected(_) => warn!("Peer already stopped."),
-            Peer::Error(_, _) => warn!("Peer broken. Handle Error first."),
         };
-    }
-}
 
-#[async_trait::async_trait]
-impl<T, K, E> ContextHandler<Start> for Peer<T, K, E>
-where
-    T: Debug + Encode + Decode + Send + Sync + Clone + 'static,
-    K: KeyExchangeScheme + Send + 'static,
-    E: Encryptor + Send + 'static,
-{
-    type Result = ();
-
-    async fn handle(&mut self, ctx: &mut Context<Self>, Start: Start) {
-        let mut dummy = Self::Disconnected(self.id().clone());
-        core::mem::swap(&mut dummy, self);
-        trace!(
-            peer = ?dummy,
-            "Starting connection and handshake"
-        );
-        if dummy.handshake().await.is_err() {
+        let (terminate_sender, mut terminate_receiver) = oneshot::channel();
+        let (post_sender, mut post_receiver) = mpsc::channel(100);
+        let ready_peer_handle = handles::PeerHandle {
+            post_sender,
+            terminate_sender,
+        };
+
+        let Ready {
+            peer_id,
+            connection:
+                Connection {
+                    read,
+                    mut write,
+                    id: connection_id,
+                },
+            cryptographer,
+        } = peer;
+
+        if connected_sender
+            .send(Connected {
+                connection_id,
+                peer_id: peer_id.clone(),
+                ready_peer_handle,
+            })
+            .is_err()
+        {
+            iroha_logger::error!(
+                "Peer is ready, but network is dropped connection sender. Terminating..."
+            );
+            let _ = terminated_sender.send(Terminated { conn_id, peer_id });
             return;
         }
-        if let Self::Ready(id, broker, mut connection, crypto) = dummy {
-            debug!(peer_addr = %id.address, "Handshake finished");
-            let connected_message = PeerMessage::<T>::Connected(id.clone(), connection.id);
-            broker.issue_send(connected_message).await;
 
-            #[allow(clippy::unwrap_used)]
-            let read: OwnedReadHalf = connection.read.take().unwrap();
+        let cryptographer = Box::pin(cryptographer);
 
-            let (sender, receiver) = oneshot::channel();
-            connection.finish_sender = Some(sender);
+        let mut message_reader = MessageReader::new(read);
 
-            // Subscribe reading stream
-            ctx.notify_with(read_connection_stream(read, receiver));
-            dummy = Self::Ready(id, broker, connection, crypto);
-            core::mem::swap(&mut dummy, self);
-        } else {
-            error!(peer = ?self, "Handshake didn't fail, but the peer is not ready");
+        // Enter peer loop
+        loop {
+            tokio::select! {
```

Comment:
I'm a big fan of this loop. I only wonder if there should be a *yield* at the top so that we don't service a single peer to the exclusion of others under high load. My understanding is that putting a yield here moves this "thread" to the back of the queue so the runner will attempt to service all the other connections before resuming this one.

Reply:
I will add yielding at the top of the loop to avoid single busy actor taking control over execution.